File: greendao-study/src/main/java/com/asura/greendao_study/utils/UIUtils.java
package com.asura.greendao_study.utils;

import android.content.Context;
import android.graphics.Color;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.view.WindowManager;
import android.widget.TextView;

import java.util.Arrays;

/**
 * 项目名：BLE App
 * 创建者：wuhj
 * 创建时间：2016/4/21 18:20
 * 描述：
 * 更新描述：
 */
public class UIUtils {

    /**
     * 判断TextView的输入是否为空
     *
     * @param textView
     * @return true/false
     */
    public static boolean isEmpty(TextView textView) {
        return TextUtils.isEmpty(textView.getText().toString().trim());
    }

    /**
     * 判断string是否为空
     *
     * @param string
     * @return true/false
     */
    public static boolean isEmpty(String string) {
        return TextUtils.isEmpty(string);
    }

    /**
     * 替代findviewById方法
     */
    public static <T extends View> T find(View view, int id) {
        return (T) view.findViewById(id);
    }

    /**
     * 根据手机的分辨率从 dp 的单位 转成为 px(像素)
     */
    public static int dip2px(Context context, float dpValue) {
        final float scale = context.getResources().getDisplayMetrics().density;
        return (int) (dpValue * scale + 0.5f);
    }

    /**
     * 将px值转换为sp值，保证文字大小不变
     *
     * @param pxValue
     * @param context （DisplayMetrics类中属性scaledDensity）
     * @return
     */
    public static int px2sp(Context context, float pxValue) {
        final float fontScale = context.getResources().getDisplayMetrics().scaledDensity;
        return (int) (pxValue / fontScale + 0.5f);
    }

    /**
     * 将sp值转换为px值，保证文字大小不变
     *
     * @param spValue
     * @param context （DisplayMetrics类中属性scaledDensity）
     * @return
     */
    public static int sp2px(Context context, float spValue) {
        final float fontScale = context.getResources().getDisplayMetrics().scaledDensity;
        return (int) (spValue * fontScale + 0.5f);
    }

    /**
     * @param context
     * @return 屏幕的宽
     */
    public static int getWindowWidth(Context context) {
        WindowManager wm = (WindowManager) context
                .getSystemService(Context.WINDOW_SERVICE);
        int width = wm.getDefaultDisplay().getWidth();
        return width;
    }

    /**
     * @param context
     * @return 屏幕的高
     */
    public static int getWindowHeight(Context context) {
        WindowManager wm = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
        int height = wm.getDefaultDisplay().getHeight();
        return height;
    }

    public static String hsvToRgb(float hue, float saturation, float value) {

        int h = (int) (hue * 6);
        float f = hue * 6 - h;
        float p = value * (1 - saturation);
        float q = value * (1 - f * saturation);
        float t = value * (1 - (1 - f) * saturation);

        switch (h) {
            case 0:
                return rgbToString(value, t, p);
            case 1:
                return rgbToString(q, value, p);
            case 2:
                return rgbToString(p, value, t);
            case 3:
                return rgbToString(p, q, value);
            case 4:
                return rgbToString(t, p, value);
            case 5:
                return rgbToString(value, p, q);
            default:
                return null;
        }
    }

    public static float[] rgbToHsb(int rgbR, int rgbG, int rgbB) {
        int[] rgb = new int[]{rgbR, rgbG, rgbB};
        Arrays.sort(rgb);
        int max = rgb[2];
        int min = rgb[0];

        float hsbB = max / 255.0f;
        float hsbS = max == 0 ? 0 : (max - min) / (float) max;

        float hsbH = 0;
        if (max == rgbR && rgbG >= rgbB) {
            hsbH = (rgbG - rgbB) * 60f / (max - min) + 0;
        } else if (max == rgbR && rgbG < rgbB) {
            hsbH = (rgbG - rgbB) * 60f / (max - min) + 360;
        } else if (max == rgbG) {
            hsbH = (rgbB - rgbR) * 60f / (max - min) + 120;
        } else if (max == rgbB) {
            hsbH = (rgbR - rgbG) * 60f / (max - min) + 240;
        }

        return new float[]{hsbH, hsbS, hsbB};
    }

    public static String rgbToString(float r, float g, float b) {
        String rs = Integer.toHexString((int) (r * 255));
        if (rs.length() == 1) {
            rs = "0" + rs;
        }
        String gs = Integer.toHexString((int) (g * 255));
        if (gs.length() == 1) {
            gs = "0" + gs;
        }
        String bs = Integer.toHexString((int) (b * 255));
        if (bs.length() == 1) {
            bs = "0" + bs;
        }


        return rs + gs + bs;
    }

    public static String rgbToColorString(int r, int g, int b) {
        String rs = Integer.toHexString(r);
        if (rs.length() == 1) {
            rs = "0" + rs;
        }
        String gs = Integer.toHexString(g);
        if (gs.length() == 1) {
            gs = "0" + gs;
        }
        String bs = Integer.toHexString(b);
        if (bs.length() == 1) {
            bs = "0" + bs;
        }
        Log.d("colortest", r + "--" + rs);

        return rs + gs + bs;
    }

    /**
     * 根据亮度的变化修改当前的颜色()
     *
     * @param color      当前颜色
     * @param brightness 变化的亮度
     * @return color值
     */
    public static int brightnessToColor(String color, float brightness) {
        return brightnessToColor(Color.parseColor(color), brightness);
    }

    /**
     * 根据亮度的变化修改当前的颜色
     *
     * @param color      当前颜色
     * @param brightness 变化的亮度
     * @return color值
     */
    public static int brightnessToColor(int color, float brightness) {
        float[] hsbValues = colorToHsb(color);
        hsbValues[2] = brightness;
        return Color.HSVToColor(hsbValues);
    }

    /**
     * 颜色转换为hsb值
     *
     * @param color 颜色
     * @return hsb值数组
     */
    public static float[] colorToHsb(int color) {
        float[] hsbValues = new float[3];
        Color.colorToHSV(color, hsbValues);
        return hsbValues;
    }
}

File: jsoup-study/src/main/java/com/asura/jsoup_study/util/UiUtils.kt
package com.asura.jsoup_study.util

import android.app.Activity
import android.content.Context
import android.graphics.drawable.Drawable
import android.os.Build
import android.support.annotation.ColorRes
import android.support.annotation.DrawableRes
import android.support.annotation.StringRes
import android.text.TextUtils
import android.util.DisplayMetrics
import android.widget.TextView
import com.asura.jsoup_study.MyApplication

/**
 * UI 工具类
 *
 * @author Created by Asura on 2018/06/07 10:32.
 */
object UiUtils {

    /**
     * 判断TextView的输入是否为空
     *
     * @param textView
     * @return true/false
     */
    fun isEmpty(textView: TextView): Boolean {
        return TextUtils.isEmpty(textView.text.toString().trim { it <= ' ' })
    }

    /**
     * 判断string是否为空
     *
     * @param string
     * @return true/false
     */
    fun isEmpty(string: String): Boolean {
        return TextUtils.isEmpty(string)
    }

    /**
     * @param src
     * @return 如果src不能转，则默认返回-1
     */
    fun stringToInt(src: String): Int {
        var res = -1
        try {
            res = Integer.parseInt(src)
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return res
    }

    fun stringToLong(src: String): Long {
        var res: Long = -1
        try {
            res = java.lang.Long.parseLong(src)
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return res
    }

    /**
     * 根据手机的分辨率从 dp 的单位 转成为 px(像素)
     */
    fun dip2px(context: Context, dpValue: Float): Int {
        val scale = context.resources.displayMetrics.density
        return (dpValue * scale + 0.5f).toInt()
    }

    fun px2dip(context: Context, pxValue: Float): Int {
        val scale = context.resources.displayMetrics.density
        return (pxValue / scale + 0.5f).toInt()
    }

    /**
     * 将px值转换为sp值，保证文字大小不变
     *
     * @param pxValue
     * @param context （DisplayMetrics类中属性scaledDensity）
     * @return
     */
    fun px2sp(context: Context, pxValue: Float): Int {
        val fontScale = context.resources.displayMetrics.scaledDensity
        return (pxValue / fontScale + 0.5f).toInt()
    }

    /**
     * 将sp值转换为px值，保证文字大小不变
     *
     * @param spValue sp
     * @param context （DisplayMetrics类中属性scaledDensity）
     */
    fun sp2px(context: Context, spValue: Float): Int {
        val fontScale = context.resources.displayMetrics.scaledDensity
        return (spValue * fontScale + 0.5f).toInt()
    }

    /**
     * @param context 上下文
     * @return 屏幕的宽
     */
    fun getWindowWidth(context: Context): Int {
        return context.resources.displayMetrics.widthPixels
    }

    /**
     * @param context 上下文
     * @return 屏幕的高
     */
    fun getWindowHeight(context: Context): Int {
        return context.resources.displayMetrics.heightPixels
    }

    /**
     * 获取drawable res文件
     */
    fun getDrawable(@DrawableRes resId: Int): Drawable {
        return MyApplication.getInstance()!!.resources.getDrawable(resId)
    }

    /**
     * 获取color
     */
    fun getColor(@ColorRes resId: Int): Int {
        return MyApplication.getInstance()!!.resources.getColor(resId)
    }

    fun getString(@StringRes resId: Int): String {
        return if (resId == -1) {
            ""
        } else MyApplication.getInstance()!!.resources.getString(resId)
    }

    fun getUrl(url: String, params: Map<String, String>?): String {
        var url = url
        // 添加url参数
        if (params != null) {
            val it = params.keys.iterator()
            var sb: StringBuffer? = null
            while (it.hasNext()) {
                val key = it.next()
                val value = params[key]
                if (sb == null) {
                    sb = StringBuffer()
                    sb.append("?")
                } else {
                    sb.append("&")
                }
                sb.append(key)
                sb.append("=")
                sb.append(value)
            }
            if (sb != null) {
                url += sb.toString()
            }
        }
        return url
    }

    /**
     * 弹出短暂(2s)Toast提示
     *
     * @param msg 提示语
     */
    fun toastShort(msg: String) {
        ToastUtils.instance.toastShort(MyApplication.getInstance()!!, msg).show()
    }

    /**
     * 弹出短暂(2s)Toast提示
     *
     * @param resId 提示语资源id
     */
    fun toastShort(@StringRes resId: Int) {
        ToastUtils.instance.toastShort(MyApplication.getInstance()!!, getString(resId)).show()
    }

    /**
     * 弹出长时间（3.5s）Toast提示
     *
     * @param msg 提示语
     */
    fun toastLong(msg: String) {
        ToastUtils.instance.toastLong(MyApplication.getInstance()!!, msg).show()
    }

    /**
     * 弹出长时间（3.5s）Toast提示
     *
     * @param resId 提示语资源id
     */
    fun toastLong(@StringRes resId: Int) {
        ToastUtils.instance.toastLong(MyApplication.getInstance()!!, getString(resId)).show()
    }

    /**
     * 判断phone导航栏是否存在
     *
     * @param activity
     * @return
     */
    fun isNavigationBarExist(activity: Activity): Boolean {
        val windowManager = activity.windowManager
        val d = windowManager.defaultDisplay

        val realDisplayMetrics = DisplayMetrics()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            d.getRealMetrics(realDisplayMetrics)
        }

        val realHeight = realDisplayMetrics.heightPixels
        val realWidth = realDisplayMetrics.widthPixels

        val displayMetrics = DisplayMetrics()
        d.getMetrics(displayMetrics)

        val displayHeight = displayMetrics.heightPixels
        val displayWidth = displayMetrics.widthPixels

        return realWidth - displayWidth > 0 || realHeight - displayHeight > 0
    }
}
